state/aws/dynamodb/dynamodb.go

```diff
@@ -0,0 +1,168 @@
+package dynamodb
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbiface"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/dapr/components-contrib/state"
+)
+
+// StateStore is a DynamoDB state store
+type StateStore struct {
+	client dynamodbiface.DynamoDBAPI
+	table  string
+}
+
+type dynamoDBMetadata struct {
+	Region       string `json:"region"`
+	AccessKey    string `json:"accessKey"`
+	SecretKey    string `json:"secretKey"`
+	SessionToken string `json:"sessionToken"`
+	Table        string `json:"table"`
+}
+
+// NewDynamoDBStateStore returns a new dynamoDB state store
+func NewDynamoDBStateStore() *StateStore {
+	return &StateStore{}
+}
+
+// Init does metadata and connection parsing
+func (d *StateStore) Init(metadata state.Metadata) error {
+	meta, err := d.getDynamoDBMetadata(metadata)
+	if err != nil {
+		return err
+	}
+
+	client, err := d.getClient(meta)
+	if err != nil {
+		return err
+	}
+
+	d.client = client
+	d.table = meta.Table
+	return nil
+}
+
+// Get retrieves a dynamoDB item
+func (d *StateStore) Get(req *state.GetRequest) (*state.GetResponse, error) {
+	input := &dynamodb.GetItemInput{
+		ConsistentRead: aws.Bool(req.Options.Consistency == state.Strong),
+		TableName:      aws.String(d.table),
+		Key: map[string]*dynamodb.AttributeValue{
+			"key": {
+				S: aws.String(req.Key),
+			},
+		},
+	}
+
+	result, err := d.client.GetItem(input)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(result.Item) == 0 {
+		return &state.GetResponse{}, nil
+	}
+
+	var output string
+	if err = dynamodbattribute.Unmarshal(result.Item["value"], &output); err != nil {
+		return nil, err
+	}
+	return &state.GetResponse{
+		Data: []byte(output),
+	}, nil
+}
+
+// Set saves a dynamoDB item
+func (d *StateStore) Set(req *state.SetRequest) error {
+	item := map[string]*dynamodb.AttributeValue{
+		"key": {
+			S: aws.String(req.Key),
+		},
+		"value": {
+			S: aws.String(fmt.Sprintf("%v", req.Value)),
+		},
+	}
+
+	input := &dynamodb.PutItemInput{
+		Item:      item,
+		TableName: &d.table,
+	}
+
+	_, e := d.client.PutItem(input)
+	return e
+}
+
+// BulkSet performs a bulk set operation
+func (d *StateStore) BulkSet(req []state.SetRequest) error {
```

Comment:
not completely happy with this implementation, I might change it to WriteBatchRequest later.

Reply:
Please, create an issue after merging.

Reply:
what worries me with batch operations done in a loop is the high probability of exceeding DynamoDB's provisioned throughput for writes. The issue is hidden in a single-element write too, because if we start calling state store too frequent from our app even with throughput autoscaling we are still going to experience 4XX responses due to insufficient provisioned throughput for a short period of time. The issue does not affect for example redis. And my question is, is there any storage supervisor that will handle retries or should the store be self-contained and do retries too.

Reply:
If the dynamo API has a bulk/batch operation that's the preferable route.

It should be assumed that retries will be handled globally by Dapr.